### ISSUE TITLE
Optimize inproc to avoid copying message

### ISSUE THREAD

**Issue by gdamore** [nanomsg/mangos#11]
Originally: https://bitbucket.org/gdamore/mangos/issue/11

The protocols that handle the header (req/rep in particular) should learn to grok a message that is passed up with the header "pre-processed". This would prevent a requirement to copy / split the headers in the inproc transport.

However:

Its unclear that this is worthwhile. We're already getting rather good performance as it stands. Tweaking this further would allow us to actually outperform nanomsg, but to what end? Inproc isn't really a real-world deployment -- especially in a Go environment where its more natural to use channels.

**Comment by lthibault** [comment]
Hello,

In certain realtime systems, there might be a real use for low-latency, in-process messaging patterns.  I was just thinking about implementing a `realtime` transport that would forego channels in favor of something like [gringo](https://github.com/textnode/gringo), so color me *very* interested.

tl;dr:  `inproc` is (sometimes) a real-world deployment scenario and I think this is worthwhile.

**Comment by binary132** [comment]
FWIW, I am using inproc and adding a TCP option in case I need to scale.  That way the same implementation is useful whether on a single node or multiple nodes.

Additionally, a channel implementation requires sharing between goroutines, while the inproc option obscures the fact that there is sharing, and requires no sharing if / when I scale.

**Comment by binary132** [comment]
@lthibault -- awesome!  I was pondering doing a performant IPC/inproc option using mmap'd files as in http://www.boost.org/doc/libs/1_47_0/doc/html/interprocess/sharedmemorybetweenprocesses.html#interprocess.sharedmemorybetweenprocesses.sharedmemory but it's a wishlist item at the moment.

**Comment by lthibault** [comment]
@binary132 Would you be open to collaborating on this issue?  I've been looking for an excuse to dig into the entrails of this project, and I'd be interested in learning about shared memory.  Where do you suggest we start?

**Comment by gdamore** [comment]
In nng (the nanomsg rewrite), the buffers have a notion of "headroom" -- meaning I preallocate space at the head of the array so that I can insert header material without having to do a reallocate or copy.

Unfortunately, go arrays have no way to do this.

**Comment by gdamore** [comment]
Spending a little more time looking at this... there are some problems with *not* copying the data, which really relate to the fact that the receiving side might want to modify or reuse the array.  In an ideal scenario we could have some sort of copy on write, but we don't get that because we let the user get directly at the slice containing the message payload.

Perhaps we could have a deferred copy with an explicit "read-only slice" routine, but at the end of the day I think the extra complexity this puts on applications is unlikely to be worth the performance increase.  (Applications that really need fast message transports should look at doing something different here.)

**Comment by gdamore** [comment]
I think its unlikely that this will ever be done in mangos.  The idea of a shared memory transport or RDMA is orthogonal.  Closing this particular issue.
